### PR TITLE
egib layers

### PR DESCRIPTION
**Pull request by gsapijaszko**
Added `egib_layers` and `egib_layer_download()`.

### CODE REVIEW

File: R/egib_layer_download.R [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+#'
+
+
+# county <- "trzebnicki"
+# # # county <- "Łomża"
+# TERYT <- NULL
+# layer <- "Osnowa"
+# outdir <- "~/Downloads/"
+
+egib_layer_download <- function(county = NULL, TERYT = NULL, layer = "dzialki", outdir = NULL, ...) {
```
Comment (kadyb) [inline R/egib_layer_download.R:40]: Could the argument `layer` take English names, i.e. `parcels` and `buildings`?

File: R/egib_layer_download.R [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+#' \url{https://www.geoportal.gov.pl/en/data/land-and-building-register-egib/}
+#' \url{https://www.geoportal.gov.pl/en/data/detailed-control-network-database-bdsog/}
+#'
+
+
+# county <- "trzebnicki"
+# # # county <- "Łomża"
+# TERYT <- NULL
+# layer <- "Osnowa"
+# outdir <- "~/Downloads/"
```
Comment (kadyb) [inline R/egib_layer_download.R:38]:
```suggestion
```

**Comment by kadyb** [comment]
The package assumes that if function returns a spatial object without saving it to disk, it takes the name `_get`. However, I wonder if it wouldn't be better if this function took multiple counties at once and downloaded them to disk, and the users could load them themselves?

File: R/egib_layer_download.R (hunk 1 above)
Comment (kadyb) [reply R/egib_layer_download.R:40]: I think we can simplify the function name to `egib_` (`download` or `get`; see comment below).

File: R/egib_layer_download.R [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+  }
+
+  if(any(grepl(layer, layers))) {
+    layer <- layers[which(grepl(layer, layers))]
+    if(length(layer) > 1L) {
+      stop(paste0("There is more than 1 layer simillar to requested: ",
+                    paste0(unlist(layer), collapse = ", "), ". Please specify."))
+    }
+    egib_url <- paste0("WFS:", egib_url)
+    output <- tryGet(sf::st_read(dsn = egib_url, layer = layer,
```
Comment (kadyb) [inline R/egib_layer_download.R:94]: Maybe replace `st_read()` with `read_sf()`, so it doesn't print information to the console?

File: R/egib_layer_download.R [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+                                 options = "CONSIDER_EPSG_AS_URN=YES", ...))
+
+    if (any(output %in% c("error", "warning"))) {
+      return(invisible("connection error"))
+    }
+
+    if(!is.null(outdir)) {
+      outfile <- paste0(outdir, "/", TERYT, ".gpkg")
+      message(paste("Writing the data to", outfile))
+      sf::st_write(output, dsn = outfile, layer = layer, append = FALSE)
```
Comment (kadyb) [inline R/egib_layer_download.R:104]: Maybe replace `st_write()` with `write_sf()`, so it doesn't print information to the console?

**Comment by kadyb** [comment]
TODO:

1. Update NEWS file.
2. Update the table in the README file.
3. Change examples and tests to download as little data as possible.

**Comment by kadyb** [comment]
Could you add yourself as contributor (`ctb`) to [DESCRIPTION](https://github.com/kadyb/rgugik/blob/master/DESCRIPTION)?

**Comment by kadyb** [comment]
If there are counties with the same name, the function returns an error:

```r
egib_layer_download(county = "grodziski")
#> Error in layers[[1]] : subscript out of bounds
```

File: R/egib_layer_download.R (hunk 2 above)
Comment (gsapijaszko) [reply R/egib_layer_download.R:38]: Done, garbage from testing.

**Comment by gsapijaszko** [comment]
> The package assumes that if function returns a spatial object without saving it to disk, it takes the name `_get`. However, I wonder if it wouldn't be better if this function took multiple counties at once and downloaded them to disk, and the users could load them themselves?

Done for multiple counties (and the same layer for all of them), `outdir` pointed out to current directory per default. A bonus: counties with same names will throw a message only and are taken for download. However, if there is more than one county AND `outdir = NULL`, only the county is taken.

File: R/egib_layer_download.R (hunk 3 above)
Comment (gsapijaszko) [reply R/egib_layer_download.R:94]: Changed.

File: R/egib_layer_download.R (hunk 4 above)
Comment (gsapijaszko) [reply R/egib_layer_download.R:104]: Replaced.

**Comment by kadyb** [comment]
I think you should update [gsapijaszko:master](https://github.com/gsapijaszko/rgugik/tree/master), then the changes will be visible in this PR.

**Comment by kadyb** [comment]
I also think that we should mention in the docs about setting timeout in case the servers turn out to be VERY slow.

https://gdal.org/en/latest/user/configoptions.html -- `GDAL_HTTP_CONNECTTIMEOUT`, `GDAL_HTTP_TIMEOUT`
https://gdal.org/en/latest/drivers/vector/wfs.html -- need to add `&Timeout` to URL in function?

**Comment by gsapijaszko** [comment]
> I also think that we should mention in the docs about setting timeout in case the servers turn out to be VERY slow.
> 
> https://gdal.org/en/latest/user/configoptions.html -- `GDAL_HTTP_CONNECTTIMEOUT`, `GDAL_HTTP_TIMEOUT` https://gdal.org/en/latest/drivers/vector/wfs.html -- need to add `&Timeout` to URL in function?

New commit is on the way, took me a while to test it. 

Regarding the timeout options - yesterday I have started a vignette for EGiB, have a Troubleshooting section. I will check the links and add a few words when needed. The issue is not with time-out, rather few of the servers are not responding at all, either they returns a garbage instead of proper XML.

G.

**Comment by kadyb** [comment]
Hi @gsapijaszko, is this ready to merge or would you like to change something?

**Comment by gsapijaszko** [comment]
> Hi @gsapijaszko, is this ready to merge or would you like to change something?

I think it's ready to be merged as it is. A vignette is still on my list and I'm looking forward for suggestion how you would like to manage:

> How to reproduce egib_layers dataset?

The options in my mind:

- [ ] add to `zzz.R` commented out script to keep it on hand?
- [ ] add it to vignette?
- [ ] other suggestion?

**Comment by kadyb** [comment]
I think the easiest option will be to paste the code to reproduce as the issue in the repository.

And it looks like the CI test doesn't work (https://github.com/kadyb/rgugik/actions/runs/11846188296):

```
GDAL Error 1: Failed to connect to swietochlowice.geoportal2.pl port 443 after 135590 ms: Connection timed out[1] 1

── Failure ('test-egib_download.R:7:1'): (code run outside of `test_that()`) ───
`status` is not an S3 object
```

**Pull request merged**